smartsim/_core/mli/infrastructure/storage/backbone_feature_store.py

```diff
@@ -352,68 +575,145 @@ def __init__(
+        :param timeout: Maximum time to wait for a single message to arrive
+        :param batch_timeout: Maximum time to wait for messages to arrive; allows
+        multiple batches to be retrieved in one call to `send`
         :returns: A list of events that pass any configured filters
         """
         if filters is None:
             filters = []
 
+        if batch_timeout is not None and batch_timeout <= 0:
+            raise ValueError("batch_timeout must be a non-zero, positive value")
+
         filter_set = {*self._global_filters, *filters}
-        messages: t.List[t.Any] = []
+        all_message_bytes: t.List[bytes] = []
 
-        # use the local timeout to override a global setting
-        start_at = time.time_ns()
+        # firehose as many messages as possible within the batch_timeout
+        start_at = time.time()
+        remaining = batch_timeout
 
-        while msg_bytes_list := self._comm_channel.recv(timeout=timeout):
+        batch_message_bytes = self._comm_channel.recv(timeout=timeout)
+        while batch_message_bytes:
             # remove any empty messages that will fail to decode
-            msg_bytes_list = [msg for msg in msg_bytes_list if msg]
+            all_message_bytes.extend(batch_message_bytes)
+            batch_message_bytes = []
+
+            # avoid getting stuck indefinitely waiting for the channel
+            elapsed = time.time() - start_at
+            remaining = batch_timeout - elapsed
 
-            msg: t.Optional[EventBase] = None
-            if msg_bytes_list:
-                for message in msg_bytes_list:
-                    msg = pickle.loads(message)
+            if remaining > 0:
+                batch_message_bytes = self._comm_channel.recv(timeout=timeout)
 
-                    if not msg:
-                        logger.warning("Unable to unpickle message")
-                        continue
+        events_received: t.List[EventBase] = []
 
-                    # ignore anything that doesn't match a filter (if one is
-                    # supplied), otherwise return everything
-                    if not filter_set or msg.category in filter_set:
-                        messages.append(msg)
+        # Timeout elapsed or no messages received - return the empty list
+        if not all_message_bytes:
+            return events_received
 
-            # avoid getting stuck indefinitely waiting for the channel
-            elapsed = (time.time_ns() - start_at) / 1000000000
-            remaining = elapsed - self._global_timeout
-            if remaining > 0:
-                logger.debug(f"Consumer batch timeout exceeded by: {abs(remaining)}")
-                break
+        for message in all_message_bytes:
+            if not message or message is None:
+                continue
+
+            event = pickle.loads(message)
+            if not event:
+                logger.warning("Unable to unpickle message")
```

Comment:
Do we need a ``continue`` here so that ``event.cateogry`` later doesn't raise an error?

Reply:
we do. added.